### PR TITLE
Helm 3: Add acceptance test suite

### LINKED ISSUE #5999: Helm 3: Add acceptance test suite
A suite of acceptance tests should be added to verify that a built version of Helm works as expected. For example, make sure a `helm install` works on multiple versions of Kubernetes.

CC @thomastaylor312

### PR DESCRIPTION
**Pull request by jdolitsky**
Fixes #5999 

Please see example test report here and click around:
https://jdolitsky.io/files/helm-acceptance-test-example/report.html

Starting with just one example test case, that the `--wait` flag will work with both Kubernetes 1.14.3 and 1.15.0. If you want to try it out, checkout acceptance_tests/README.md

Once merged, will follow-up with PR to add this to CI

### CODE REVIEW

**Review by thomastaylor312 — COMMENTED** [review]
Overall this looks really nice. I have to admit, I kinda like the plain English test descriptions Robot provides. With that said, I do have one question that I think we should have good answers for before we merge: Are we ok with requiring Python as a dependency for running these tests? Could they reasonably be written in Go with a different framework?

I don't think those questions are blockers by the way. I just want to make sure we have good answers to point to if anyone has any questions in the future

**Comment by jdolitsky** [comment]
@thomastaylor312 thanks for taking a look

> Are we ok with requiring Python as a dependency for running these tests? Could they reasonably be written in Go with a different framework?

I'm personally OK with it 😄 Acceptance testing in and of itself should be agnostic to the language that the app being tested is written in.

ChartMuseum uses two separate Python-based frameworks for both acceptance testing and load testing (robot and locust). Robot is picked based on ease of use, and well as advanced HTML reporting features. Most AC frameworks have similar easy-to-use, english definitions, but none of them have this robust HTML report baked in. Terminal output is simply not enough. I encourage you to drill down in the report from the link above, you can see each individual command run and its output as part of a test.

I understand the desire to only require Go for a Go-based project, but none seem to compare to the capability/maturity of Robot. The best I've found is [godog](https://github.com/DATA-DOG/godog/), which essentially outputs as cucumber. In order to turn this into an HTML report, you need something like [cucumber-html-reporter](https://github.com/gkushang/cucumber-html-reporter) (which requires node.js/npm to install and use).

So, I'm confident it's the best option, but I'm certainly biased for having done the work already. Maybe a middle ground would be to run Robot in a Docker container instead? There is a public image available [here](https://hub.docker.com/r/robotframework/rfdocker). This would shift "you need Python3" to "you need Docker".

File: acceptance_tests/testdata/charts/nginx/templates/pvc.yaml
Comment (jdolitsky) [reply acceptance_tests/testdata/charts/nginx/templates/pvc.yaml]: done

File: acceptance_tests/testdata/charts/nginx/templates/deployment.yaml
Comment (jdolitsky) [reply acceptance_tests/testdata/charts/nginx/templates/deployment.yaml]: done

File: acceptance_tests/lib/Kubectl.py
Comment (jdolitsky) [reply acceptance_tests/lib/Kubectl.py]: That's a good idea. I'd like to open another ticket to migrate to use that instead once this is merged

Comment (thomastaylor312) [reply acceptance_tests/lib/Kubectl.py]: That sounds fine to me!

**Comment by thomastaylor312** [comment]
@jdolitsky ok that sounds like great reasoning to me! I think for local development it might be worth having a docker container, but we can add that later if it looks like it would be useful to anyone contributing. 

Let's get at least one more of the other maintainers to take a look at this and it should be good to go

**Review by thomastaylor312 — APPROVED** [review]

File: acceptance_tests/README.md
Comment (bacongobbler) [inline acceptance_tests/README.md]: `open` is a command that only exists for MacOS. Can we instead recommend others to open this file in a browser?

File: acceptance_tests/README.md
Comment (bacongobbler) [inline acceptance_tests/README.md]: typo: run*

File: acceptance_tests/README.md
Comment (bacongobbler) [inline acceptance_tests/README.md]: is there a basic system requirement to run these tests? e.g. what are the specs of your PC running this test suite?

**Comment by bacongobbler** [comment]
Given that these tests are independent of Helm core, I wonder if it'd make sense to spin this off as a separate project similar to [chart-testing](https://github.com/helm/chart-testing). This smells a lot like the [Linux Test Project](https://linux-test-project.github.io/), which helped test and validate the Linux kernel for compatibility but was ran by a separate group than the linux core maintainers, but it didn't make any assumptions about the internal APIs.

What do you think about spinning this off as a separate initiative? That way the development of core and the test suite can move independent of each other.

**Comment by bacongobbler** [comment]
another parallel project similar in nature would be [prow](https://github.com/kubernetes/test-infra).

**Comment by jdolitsky** [comment]
@bacongobbler I'm happy to move it to a new project, I suppose it can just run against whatever "helm" is in PATH (vs. bin/helm)

**Comment by jdolitsky** [comment]
I can go ahead and start it up, how does "helm/acceptance-testing" sound? Open to whatever

**Comment by bacongobbler** [comment]
helm/acceptance-testing sounds good to me!

It'd be nice to send an email off to the mailing list first to officially kick off the process and gather lazy consensus. That way others have an opportunity to chime in here. Would you mind going ahead with that before creating the repository?

**Comment by mattfarina** [comment]
I agree with @bacongobbler on the new repo bit. His examples are the same ones I would have pulled out.

@jdolitsky note, creating a new repo does require it to go through the org maintainers. Something to the mailing list with a lazy consensus voting decision should be good.

**Comment by jdolitsky** [comment]
@mattfarina @bacongobbler I'll go ahead and send out the email. Thanks! Closing

**Comment by marckhouzam** [comment]
@jdolitsky @bacongobbler @mattfarina  Would there be interest in hooking the acceptance tests into the main helm makefile? During development, I was thinking it would be easier to simply run
```
make acceptance
```
instead of
```
cd ../accetance-testing
make accetance
cd ../helm
```
Granted that running all the acceptance tests often during development may not be a very common case, but since there will be subsets of tests within the acceptance-testing repo (like tests for completion), I was thinking the ability to trigger them from the helm repo would be nice.

If you're open to the idea, I can post a PR for comments.

**Pull request closed without merging**
